image/store.go

```diff
@@ -68,21 +68,23 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
 
 func (is *store) restore() error {
 	err := is.fs.Walk(func(dgst digest.Digest) error {
+		logger := logrus.WithField("digest", dgst)
 		img, err := is.Get(ID(dgst))
 		if err != nil {
-			logrus.Errorf("invalid image %v, %v", dgst, err)
+			logger.WithError(err).Error("invalid image")
 			return nil
 		}
 		var l layer.Layer
 		if chainID := img.RootFS.ChainID(); chainID != "" {
+			logger = logger.WithField("chainID", chainID).WithField("os", img.OperatingSystem())
```

Comment:
Nit: chaining two `(*logrus.Entry).WithField` calls is surprisingly inefficient compared to `.WithFields` [as it is implemented in terms of the latter](https://github.com/sirupsen/logrus/blob/f8bf7650dccb756cea26edaf9217aab85500fe07/entry.go#L119-L122) and deep-copies the `Entry` on each call.

```suggestion
			logger = logger.WithFields(logrus.Fields("chainID": chainID, "os": img.OperatingSystem()))
```

<details><summary>Benchmarks</summary>

```
goos: darwin
goarch: amd64
pkg: github.com/corhere/logbench
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkWithField-12     	   10000	    344272 ns/op	  338640 B/op	      15 allocs/op
BenchmarkWithFields-12    	 1000000	      2372 ns/op	     566 B/op	       3 allocs/op
PASS
```

```go
package logbench

import (
	"io"
	"strconv"
	"testing"

	"github.com/sirupsen/logrus"
)

func init() {
	logrus.SetOutput(io.Discard)
}

func BenchmarkWithField(b *testing.B) {
	entry := logrus.NewEntry(logrus.StandardLogger())
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		entry = entry.WithField(strconv.Itoa(i), i)
	}
	entry.Info("hi")
}

func BenchmarkWithFields(b *testing.B) {
	entry := logrus.NewEntry(logrus.StandardLogger())
	fields := logrus.Fields{}
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		fields[strconv.Itoa(i)] = i
	}
	entry.WithFields(fields).Info("hi")
}
```

</details>

Reply:
That's actually a good point here; this code is ran on `ImageStore.restore()`, which iterates over all images (which could be "many"). The `logger` was added to make the code a bit more DRY (see discussion with @vvoland above https://github.com/moby/moby/pull/44426#discussion_r1017921610), but looking now, I don't think that's a good idea _for this specific case_;

- the logger is only used in error conditions (the code is treats these as non-fatal, but logs them instead)
- but the `logger` would be constructed unconditionally, which means a lot of allocations for something that (in most cases) wouldn't be used.

Let me change it back to instantiate the `logger.WithField(s)` only if it's gonna be used. It's less "pretty" but probably good for this case.